### PR TITLE
Add Custom Sites to the Facebook Container

### LINKED ISSUE #422: In-content panel enabling users to "allow FB to track me on this site"
We have a design here:
https://mozilla.invisionapp.com/d/main#/console/17203644/365230778/preview
(Copy to be updated.)

### LINKED ISSUE #550: Add ability to edit allowed sites through options UI
The users should be able to view and remove the allowed sites through the options UI. It's currently possible using the panel, but few users run with the Facebook Container browser action visible on their toolbar. Having that ability also built into the options UI would make it easier for users to find.

### PR DESCRIPTION
**Pull request by maxxcrawford**
This PR adds buttons to the doorhanger panel for users to add/remove the active page/domain to the Facebook container. 

Fixes #422 
Fixes #550  

**PR To-do List**
- [x] Update onboarding panel copy/images
- [x] Add tooltip text
- [x] Fix issue where "Add this site" shows up on default FB URLs (from [FACEBOOK_DOMAINS](https://github.com/mozilla/contain-facebook/blob/master/src/background.js#L7-L23)) 
- [x] Add test directions to this PR 

## Testing Steps

Use the following directions to verify this PR works as is. 

1. **Run the add-on.** _(This will open up the add-on in a temporary new browser)_

```
$ web-ext run -s src/ 
```

2. **Go to the following sites:** 

- Non-Facebook tracking site: 
  - https://duckduckgo.com/ (There should be no Facebook trackers blocked on this site) 
- Facebook domain 
  - https://facebook.com/
-  Facebook tracking site
  - https://maxxcrawford.github.io/social-implementation-examples/facebook/login/
  - https://buzzfeed.com/

3. Once on the site, verify that trackers are detected as expected. 

4. Click on the Facebook Container add-on icon within the browser chrome to trigger the panel:   image

5. Attempt to add that site to the Facebook Container: 

- Non-Facebook tracking site and Facebook tracking site should be able be to added by clicking the following button:
image

Once this button is clicked, the expected action is for the tab to close and reopen the same URL **INSIDE** the Facebook Container. 

6. Once added, attempt to remove the newly added site from the Facebook Container: 
- Confirm you're in the Facebook Container
image

- Click on the Facebook Container add-on icon within the browser chrome to trigger the panel: 
image

  - On default Facebook Container sites (like Facebook.com and Instagram.com), these sites cannot be removed. 
  - On non-default sites, it should close and reload the page **OUTSIDE** of the Facebook Container. 

## Screenshots

No trackers detected panel (outside of FB Container): 
image

Trackers detected panel (outside of FB Container): 
image

Custom site added to FB Container
image

On default Facebook domain: 
image

On default Facebook domain (hover tooltip): 
image

In-page UI Panel (Revised Copy + Image): 
image

New Panel for Adding Custom Site: 
image

### CODE REVIEW

File: src/panel.css [hunk 1]
```diff
@@ -174,6 +174,38 @@ a:hover {
...
+
+#button-remove-site-tooltip {
+  display: none;
+  padding: 5px 10px 10px;
+  width: 250px;
+  max-width: 250px;
+  z-index: 9999;
+  position: absolute;
+  left: 1rem;
+  bottom: 2em;
```
Comment (maxxcrawford) [inline src/panel.css]: Note that this differs from the InVision mockup as the panel has `overflow:hidden` on it, not allowing for the tooltip to be visible outside of its' parent container. 

InVision mockup: 
image

cc/ @e-pang

**Comment by maxxcrawford** [comment]
@e-pang Note that I updated the CSS to bold the first sentence in each panel, per the InVision mockups: 

image

**Comment by ddurst** [comment]
I'm struggling with the strings in this a bit, but maybe it's just me.

When we talk about "sites allowed in FBC," it sounds to me like we're putting other things in that container. Phrased like that, it sounds like we're _just_ lumping sites together, or putting things in the same container. Maybe this is a technical detail (maybe it's not), but for the user, the potential ramifications aren't clearly communicated.

To me, "sites allowed" means that Facebook can access your data (I'm deliberately oversimplifying) -- if we're trying to make it clear that FB is contained/isolated, what "allowing" a site means to me is that the user is allowing FB to potentially access your data through/on any site you "allow."

I'd expect this to be more like "Don't block Facebook trackers on these sites" or something like that -- not to sound too opinionated, but not to softball it either.

**Comment by ddurst** [comment]
(I don't know that we can change this, but in the future perhaps we could/should.)

File: src/_locales/en/messages.json [hunk 2]
```diff
@@ -78,21 +78,33 @@
...
-      "description": "This is the subhead for the third onboarding panel."
+    "onboarding4-subhead": {
+      "message": "Outside of Facebook",
+      "description": "This is the subhead for the fourth onboarding panel."
     },
-    "onboarding3-p1": {
-      "message": "Buttons that let you like or share things to Facebook or let you log in to websites using your Facebook account contain trackers. Facebook Container will block these trackers and display the fence icon to show you where trackers were removed.",
-      "description": "This is the first paragraph of the third onboarding panel."
+    "onboarding4-p1": {
+      "message": "Even when you're not on Facebook, Facebook can track you online. Buttons that let you post things to Facebook or log in to websites using your Facebook account contain trackers. Facebook Container will block these trackers and display the fence icon to show you where trackers were removed.",
```
Comment (flodolo) [inline src/_locales/en/messages.json]: Use proper apostrophes `’`.

File: src/panel.js [hunk 3]
```diff
@@ -435,6 +529,27 @@ const buildAllowedSitesPanel = async(panelId) => {
   getLocalizedStrings();
 };
 
+const removeSiteFromContainer = async () => {
+
+  const thisHostname = await getActiveHostname();
+
+  await browser.runtime.sendMessage( {removeDomain: thisHostname} );
+  browser.tabs.reload();
```
Comment (kendallcorner) [inline src/panel.js]: and here.

File: src/panel.js [hunk 4]
```diff
@@ -100,13 +103,43 @@ const addFullWidthButton = (fragment, listenerClass) => {
   return button;
 };
 
+const addTooltip = (wrapper, stringId) => {
+  const div = document.createElement("div");
+  div["id"] = stringId;
```
Comment (kendallcorner) [inline src/panel.js]: I'm assuming this is just making a <div id=[stringId]> ? I didn't know you could do this. Cool.

**Comment by kendallcorner** [comment]
Also, we've already discussed a little that this has potential to conflict with MAC. 

When I have a site assigned in MAC, and then try to add it to fbc, it doesn't take. It just keeps getting opened in my MAC container. I believe we can use the messaging already setup (`sendJailedDomainsToMAC()`) to send these new sites to MAC as well.

File: src/panel.js [hunk 5]
```diff
@@ -100,13 +103,43 @@ const addFullWidthButton = (fragment, listenerClass) => {
...
+
+const isSiteInContainer = async(panelId) => {
+
+  if (panelId === "on-facebook") {
+    // Site is on default FBC domain. Show the "remove site" button, in a disabled state.
+    return true;
+  }
+
+  const addedSitesList = await browser.runtime.sendMessage("what-sites-are-added");
+  const hostname = await getActiveHostname();
```
Comment (groovecoder) [inline src/panel.js]: nit: `const activeTabHostname = await getActiveTabHostname();`

File: src/panel.js [hunk 6]
```diff
@@ -100,13 +103,43 @@ const addFullWidthButton = (fragment, listenerClass) => {
...
+
+  if (addedSitesList.includes(hostname)) {
+    return true;
+  } else {
```
Comment (groovecoder) [inline src/panel.js]: nit: drop the `else` - `return true` will exit the function so the `else` block is not needed. I trim a lot of `else` blocks out of my code just for readability.

File: src/panel.js [hunk 7]
```diff
@@ -291,22 +360,39 @@ const buildPanel = async(panelId) => {
     addLearnMoreLink(contentWrapper);
   }
 
-  addLearnHowFBCWorksButton(fragment);
+  await addLearnHowFBCWorksButton(fragment);
+
+  if (panelId === "no-trackers") {
+    await addCustomSiteButton(fragment, panelId);
+  }
```
Comment (groovecoder) [inline src/panel.js]: This is called below on line 369 whether or not `panelId === "no-trackers"` - so do we need this if block?

File: src/panel.js [hunk 8]
```diff
@@ -116,6 +149,37 @@ const addLearnHowFBCWorksButton = async (fragment) => {
   addSpan(button, "sites-added-subhead");
 };
 
+const addCustomSiteButton = async (fragment, panelId) => {
+  const shouldShowRemoveSiteButton = await isSiteInContainer(panelId);
+  let button;
```
Comment (groovecoder) [inline src/panel.js]: We don't return this value from the function, so we can probably drop it?

```js
 const addCustomSiteButton = async (fragment, panelId) => {
   const shouldShowRemoveSiteButton = await isSiteInContainer(panelId);
   if (shouldShowRemoveSiteButton) {
     addFullWidthButton(fragment, "remove-site-from-container");
     addSpan(button, "button-remove-site");
     addTooltip(button, "button-remove-site-tooltip");
     return;
   }
   addFullWidthButton(fragment, "add-site-to-container");
   addSpan(button, "button-allow-site");
}
```

File: src/panel.js [hunk 9]
```diff
@@ -116,6 +149,37 @@ const addLearnHowFBCWorksButton = async (fragment) => {
...
+
+  if (panelId === "on-facebook") {
+    const removeSiteFromContainerLink = document.querySelector(".remove-site-from-container");
+    removeSiteFromContainerLink.classList.add("disabled-button");
+    return;
+  }
+
+  if (shouldShowRemoveSiteButton) {
+    const removeSiteFromContainerLink = document.querySelector(".remove-site-from-container");
+    removeSiteFromContainerLink.addEventListener("click", async () => removeSiteFromContainer());
```
Comment (groovecoder) [inline src/panel.js]: nit: can remove the `else` block here with a `return;` early statement, like in the above `if (panelId === "on-facebook") {` block.

File: src/panel.js [hunk 10]
```diff
@@ -188,14 +252,19 @@ const handleOnboardingClicks = async(e, res) => {
   }
 
   if (buttonId === "btn-back") {
-    buildOnboardingPanel(res - 1);
+    if (res === 4) {
+      // This accounts for the skipped Panel #3
+      buildOnboardingPanel(2);
+    } else {
```
Comment (groovecoder) [inline src/panel.js]: nit: remove `else` block with an early `return;`

File: src/panel.js [hunk 11]
```diff
@@ -256,7 +325,7 @@ document.addEventListener("DOMContentLoaded", async () => {
 
   const onboarding = (currentPanel.includes("onboarding"));
   if (!onboarding) {
-    return buildPanel(currentPanel);
+    return await buildPanel(currentPanel);
```
Comment (groovecoder) [inline src/panel.js]: nit: [don't really need to `return await`](https://eslint.org/docs/rules/no-return-await)

File: src/_locales/en/messages.json (hunk 2 above)
Comment (maxxcrawford) [reply src/_locales/en/messages.json]: Updated!

File: src/panel.js (hunk 4 above)
Comment (maxxcrawford) [reply src/panel.js]: Yup. We have to fabricate all the HTML we inject in the extension (Both here in the content script and also in the panel).

File: src/panel.js (hunk 8 above)
Comment (maxxcrawford) [reply src/panel.js]: The `button` is passed as an argument in both the `addSpan` and `addTooltip` functions. I think we need to keep it.

File: src/panel.js (hunk 7 above)
Comment (maxxcrawford) [reply src/panel.js]: Yes – It's a quirk for how its' rendered in the panel. I tried different combinations/ordering of these.

File: src/panel.js [hunk 12]
```diff
@@ -435,6 +529,27 @@ const buildAllowedSitesPanel = async(panelId) => {
...
+
+const addSiteToContainer = async () => {
+
+  const thisHostname = await getActiveHostname();
+
+  const fbcStorage = await browser.storage.local.get();
+  fbcStorage.domainsAddedToFacebookContainer.push(thisHostname);
+  await browser.storage.local.set({"domainsAddedToFacebookContainer": fbcStorage.domainsAddedToFacebookContainer});
+  //await browser.runtime.sendMessage(thisHostname);
+  browser.tabs.reload();
```
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply src/panel.js]: This is a great point. I'm going to file this as a separate issue, as it also applies to the current UI when adding a site via the in-page sign-in dialog.

[the remark this answers is not in the document]
Comment (maxxcrawford) [reply src/panel.js]: Created. https://github.com/mozilla/contain-facebook/issues/578

**Pull request merged**